lib/printer.ts

```diff
@@ -2147,17 +2147,13 @@ function genericPrintNoParens(path: any, options: any, print: any) {
         ]);
 
     case "TSAsExpression": {
-        var withParens = n.extra && n.extra.parenthesized === true;
-        if (withParens) parts.push("(");
         const expression = path.call(print, "expression");
-        const expressionType = path.getValue().expression.type;
-        const needParens = expressionType === "ArrowFunctionExpression" || expressionType === "FunctionExpression";
```

Comment:
My testing indicates that `FunctionExpressions` do not need parens when they are the `expression` of a `TSAsExpression`. That is to say, `(function() {} as () => number)` is parseable. I also verified the failing case: `(() => {} as () => number)` is not parseable, and requires the arrow function to be wrapped in parens.